### ISSUE TITLE
AutoEncoder fail ?

### ISSUE THREAD

**Issue by cbalint13** [jdeng/rbm-mnist#1]
With default sample of {img_size,500,30,500,img_size} autoencoder goes like:
- Pretrain goes as expected 784]-500-50-500-[784:
  10000 samples read
  New RBM 784 -> 500
  New RBM 500 -> 30
  New RBM 30 -> 500
  New RBM 500 -> 784
  layer: 0, epoch: 0, batch: 10, error: 0.0289122, energy: -3.33495
  .....
  So pre-training here goes for: layer0, layer1 and no more.
- At final train stage it says:
  Mirroring RBM: 2 30 -> 500
  Mirroring RBM: 3 500 -> 784
  epoch: 0, batch: 0, samples: 500

Does it need to re-mirror ?

And whole process goes with repeating "x3 failed"
.......
x3 = 7.51421e-16 failed
applying new weights to 0+
epoch: 0, batch: 8, samples: 500
evaluating: cost=0, error=10.9042
evaluating: cost=0, error=10.6183
evaluating: cost=0, error=10.8329
.....

**Comment by jdeng** [comment]
It wasn't tested yet.. The x3 failed means CG is not working as expected. I'll take a look later. Thanks for testing.

**Comment by cbalint13** [comment]
- Not sure, but why CG train step need to re-mirror once more the initial NN layers !? The original starting configuration is already a mirrored thing, I cant see need to re-mirror (append) one more half. But I might be wrong.

**Comment by jdeng** [comment]
The mirroring is only done once after pretrain. Check out pretrain().

**Comment by cbalint13** [comment]
I had no luck in figure it out. Would be great if would be fixed.

**Comment by jdeng** [comment]
Did you tried other network topology and change the training parameter (e.g number of batches etc).

**Comment by cbalint13** [comment]
- Yes I tried more parameters and different layouts. However if you look at the output image there is something wrong with the layout, can see that RBM is not mirrored 1:1, something is wrong with the second half of RBM.
- Here is the output with 768]-500-300-30-300-500-[768: http://openrisc.rdsor.ro/rbm-169.png
- Here is the output with 768]-500-30-500-[768 http://openrisc.rdsor.ro/rbm-111.png
- Also on other side does "rbm->type_ = RBM::Type::LINEAR" make all neurons linear ? Because an auto-encoder should have input, output and bottleneck linear and rest of layers should be sigmoid.
